Gru unit recurrent

skaae opened the pull request:
- Add gru units
- Add gru tests
- fix bug in unroll_scan when fn did not return a list
- modify the unroll_scan_bck tests to run without mask
- add gru layer to docs
- While correcting the GRU docs i discovered a few erros in the docs that also applied to the other recurrent layers. These changes are included here

Note: We should cosider renamed either self.unroll_scan or utils.unroll_scan

skaae committed:
add gru layer

Adds gru layer

skaae committed:
add GRU layer tests + modify test_****_unroll_scan_bck to run without
mask.

skaae committed:
fix bug in unroll scan when fn did not return list

skaae committed:
add GRULayer to doc

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -8,10 +8,12 @@
 The following recurrent layers are implemented:
 
 .. autosummary::
+    :nosignatures:
```
What's this?

skaae replied:
It renders the function table without signaures. You can also see updates.py in lasagne. 
Sphinx truncates the signatues so they are useless anyway.

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -757,3 +762,306 @@ def step_masked(input_dot_W_n, mask_n, cell_previous, hid_previous,
+class GRULayer(Layer):
+    """Gated Recurrent Layer (GRU)
+
+    Layer with gated recurrent units implemented as in [1]_, [2]_.
```
I think they actually propose slightly different GRU equations, see footnote 1 in [2]; it's worth mentioning which one we are using.

I think we should have the update equations here, and for the other layers too, but you can ignore that here because we need to update all layers.

skaae replied:
Ok. I can add the equation that i use here?

skaae replied:
I would like to do this:

``` Python
    Notes
    -----
    The GRU units is implemented slightly differently in [1]_ and in [2]_.
    We use the following equation to calculate the hidden update:

    .. math:: \hat{h}^j_t = tanh(W\mathbf{x}_t + \mathbf{r}_t \odot (U\mathbf{h_{t-1}}))^j

    An alternative formulation is:

    .. math:: \hat{h}^j_t = tanh(W\mathbf{x}_t + \odot U(\mathbf{r}_t\mathbf{h_{t-1}}))^j

    We use the first formulation because it allows us to do all matrix
    operations in a single dot product.
```

Do you know howi break the math in the doc string its too wide?

craffel replied:
`.. math::` starts a math block, so you can just do

```
.. math::
  \hat{h}^j_t = tanh(W\mathbf{x}_t
                     + \mathbf{r}_t \odot (U\mathbf{h_{t-1}}))^j
```

or whatever.  The block ends with a double line break.

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -757,3 +762,306 @@ def step_masked(input_dot_W_n, mask_n, cell_previous, hid_previous,
+        The layer feeding into this layer, or the expected input shape.
+    num_units : int
+        Number of hidden units.
+    W_in_to_resetgate : function or np.ndarray or theano.shared
```
Each argument needs documentation.

skaae replied:
Can you elaborate? :)

craffel replied:
You have arguments to the function without any documentation, you need to do e.g.

```
W_in_to_resetgate : Theano shared variable, numpy array or callable
    Initializer for input-to-reset gate weight matrix.
```

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -757,3 +762,306 @@ def step_masked(input_dot_W_n, mask_n, cell_previous, hid_previous,
+    W_in_to_cell : function or np.ndarray or theano.shared
+    W_hid_to_cell : function or np.ndarray or theano.shared
+    b_cell : function or np.ndarray or theano.shared
+    nonlinearity_resetgate : function
```
Instead of `function`, it should be `callable or None` (need to fix this in all the layers, so you can ignore that here).

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -757,3 +762,306 @@ def step_masked(input_dot_W_n, mask_n, cell_previous, hid_previous,
+            [self.W_in_to_resetgate, self.W_in_to_updategate,
+             self.W_in_to_cell], axis=1)
+
+        # Same for hidden to gate weight matrices
```
Pasted comment doesn't make sense here.

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -757,3 +762,306 @@ def step_masked(input_dot_W_n, mask_n, cell_previous, hid_previous,
+            [self.W_hid_to_resetgate, self.W_hid_to_updategate,
+             self.W_hid_to_cell], axis=1)
+
+        # Stack gate biases into a (4*num_units) vector
```
It's 3*num_units in the GRU.

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -757,3 +762,306 @@ def step_masked(input_dot_W_n, mask_n, cell_previous, hid_previous,
+        # Initialize hidden state
+        if isinstance(hid_init, T.TensorVariable):
+            if hid_init.ndim != 2:
+                raise ValueError("When a tensor hid_init should be a matrix")
```
For all of these value errors, "tensor" should be changed to "TensorVariable" (a matrix is always a tensor ;) )  Needs to be changed in all layers.

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -757,3 +762,306 @@ def step_masked(input_dot_W_n, mask_n, cell_previous, hid_previous,
+            `mask[i, j] = 0` when `j > (length of sequence i)`.
+        """
+
+        # Treat all layers after the first as flattened feature dimensions
```
This should say "Treat all dimensions after the second"; this is an error in all of the layers.

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -757,3 +762,306 @@ def step_masked(input_dot_W_n, mask_n, cell_previous, hid_previous,
+        # precompute inputs*W and dimshuffle
+        # Input is provided as (n_batch, n_time_steps, n_features)
+        # W_in is (n_features, 3*num_units). input dot W is then
+        # (n_batch, n_time_steps, 3*num_units). Because scan iterate over the
```
"Because scan iterate.." looks like this got erroneously copy/pasted here.

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -757,3 +762,306 @@ def step_masked(input_dot_W_n, mask_n, cell_previous, hid_previous,
+
+        # Create single recurrent computation step function
+        # input_dot_W_n is the n'th row of the input dot W multiplication
+        def step(input_dot_w_n, hid_previous, W_hid_to_gates):
```
Do you mind making it `input_dot_W_n` for consistency, even though your editor doesn't like it?

skaae replied:
No. I forgot to change it.

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -757,3 +762,306 @@ def step_masked(input_dot_W_n, mask_n, cell_previous, hid_previous,
+            resetgate = self.nonlinearity_resetgate(resetgate)
+            updategate = self.nonlinearity_updategate(updategate)
+
+            cell = cell_in + resetgate*cell_hid
```
`cell` is sort of a strange name here -- based on the notation in [2], it seems like a better name would be something like `hidden_update` or something.  What do you think?

skaae replied:
I agree.`hidden_update` is fine by me.

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -757,3 +762,306 @@ def step_masked(input_dot_W_n, mask_n, cell_previous, hid_previous,
+            # repeat num_batch times
+            hid_init = T.dot(T.ones((num_batch, 1)), self.hid_init)
+
+        # Scan op iterates over first dimension of input and repeatedly
```
Please remove duplicated comment.

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -757,3 +762,306 @@ def step_masked(input_dot_W_n, mask_n, cell_previous, hid_previous,
+        self.b_cell = self.add_param(
+            b_cell, (num_units,), name="b_cell", regularizable=False)
+
+        self.W_in_to_gates = T.concatenate(
```
`to_gates` doesn't make a lot of sense for `W_in_to_cell`.  I would be fine calling this simply `W_in` (and `W_hid` below) or `W_in_stacked` or something.  This is true for all of the layers, so you can ignore it here...

skaae replied:
yes. I think `W_in_stacked` and `W_hid_stacked` is fine?

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -757,3 +762,306 @@ def step_masked(input_dot_W_n, mask_n, cell_previous, hid_previous,
+            cell_in = slice_w(input_dot_w_n, 2)
+            cell_hid = slice_w(hid_input, 2)
+
+            resetgate = self.nonlinearity_resetgate(resetgate)
```
Let's move these to line 989, above `cell_in` and `cell_hid`.

craffel commented:
Just made a bunch of mostly superficial comments, which I can change later if you want.  Many of them are issues with all of the layers I noticed just because I was scrutinizing this layer class.  Let me know if you want to change them, otherwise I will merge and I can make the changes myself.

skaae commented:
I can fix it layer tonight. Should i apply your changes to all layers where it applies? It's no big deal while im add it, but it does mess up the PR a little bit.

craffel commented:
Either you should do it to all layers here, or I should just do it to the rest of the layers myself and you can rebase and fix the GRU only.  What do you want to do?

skaae committed:
update to use callable instead of function

- rename nonlinearity_cell to nonlinearity_hid

skaae committed:
doc and var renaming

skaae commented:
Adressed your comments. Im not 100% satisfied with the names `hidden_update_in` and `hidden_update_in`. 

Do you know how to write math where the latex span several lines in the doc?

skaae committed:
fix pep8 error

craffel commented on lasagne/layers/recurrent.py:
```diff
@@ -757,3 +773,310 @@ def step_masked(input_dot_W_n, mask_n, cell_previous, hid_previous,
+        self.unroll_scan = unroll_scan
+
+        # Input dimensionality is the output dimensionality of the input layer
+        num_batch = self.input_shape[0]
```
`num_batch` isn't used.

skaae committed:
remove unused args in GRU

skaae committed:
add note about hidden candidate activation

skaae committed:
add missing parameter docstrings

skaae commented:
If you don't mind i'll remove the math from the LSTM weight docstrings?

craffel commented:
As long as you replace them with sufficiently descriptive explanations, that's fine!

skaae commented:
Is `Initializer for input-to-reset gate weight matrix`fine?

craffel commented:
Well, all of the other layers have docs for their parameter initializers which look like e.g.

```
        An initializer for the weights of the layer. If a shared variable or a
        numpy array is provided the shape should  be (num_inputs, num_units).
        See :meth:`Layer.create_param` for more information.
```

If we made it that long for every parameter, we'd have a stupid huge docstring with a lot of repetition... but, I do think it's worth mentioning the required shape.  For the rest of it, maybe at the top of the docstring we can say something like

```
    All parameter initializers should be a callable, a numpy array of the initial 
    values, or a Theano shared variable.  See :meth:`Layer.create_param`
    for more information.
```

I can handle this stuff if you want!  I see it as the last clean-up we're going to need to do before submitting a PR to the main repo.

skaae commented:
Yes go ahead. I think i discussed the length of weight and bias descriptions with either sander or jan. They suggested shorter descriptions as far as i remember.

skaae commented:
btw we need a test for line 111 to avoid ruining the 100% coverage badge :)

skaae committed:
cleanup tests

- Use eval instead of compiling functions
- delete wrongly placed comments
- add one missing assert

skaae commented:
I cleaned up the tests. I'll be gone for the next few days so i uploaded the cleaned tests for reccurrent/lstm/gru here. I hope that is allright. 

The changes are mostly variable naming and eval instead of compiling functions. Finally it adds a missing assert.

craffel commented:
Ok, thanks, merging!
